packages/server-wallet/src/protocols/__test__/defund-channel.test.ts

```diff
@@ -0,0 +1,175 @@
+
+  await channelDefunder.crank(obj, response);
+
+  const reloadedObjective = await store.getObjective(obj.objectiveId);
+  expect(reloadedObjective.status).toEqual('succeeded');
+  expect(concludeSpy).toHaveBeenCalled();
+});
+
+test('when the channel does not exist it should fail the objective', async () => {
+  const chainService = new MockChainService();
+  const channelDefunder = ChannelDefunder.create(store, chainService, logger, timingMetrics);
+  const c = channel();
+
+  const obj = createPendingObjective(c.channelId);
+  await knex.transaction(tx => ObjectiveModel.insert(obj, tx));
+  const response = WalletResponse.initialize();
+
+  await channelDefunder.crank(obj, response);
+
+  const reloadedObjective = await store.getObjective(obj.objectiveId);
+  expect(reloadedObjective.status).toEqual('failed');
+});
+test('when using non-direct funding it fails', async () => {
+  const chainService = new MockChainService();
+  const channelDefunder = ChannelDefunder.create(store, chainService, logger, timingMetrics);
+
+  const c = channel({fundingStrategy: 'Fake'});
+
+  await Channel.query(knex).withGraphFetched('signingWallet').insert(c);
+
+  const obj = createPendingObjective(c.channelId);
+  await knex.transaction(tx => store.ensureObjective(obj, tx));
+  const response = WalletResponse.initialize();
+
+  await channelDefunder.crank(obj, response);
+
+  const reloadedObjective = await store.getObjective(obj.objectiveId);
+  expect(reloadedObjective.status).toEqual('failed');
+});
+
+test('when there is an active challenge and no conclusion proof it should do nothing', async () => {
+  const chainService = new MockChainService();
+  const channelDefunder = ChannelDefunder.create(store, chainService, logger, timingMetrics);
+
+  const c = channel();
+  const challengeState = {
+    ...c.channelConstants,
+    ...stateVars(),
+  };
+  await Channel.query(knex).withGraphFetched('signingWallet').insert(c);
+  await ChallengeStatus.insertChallengeStatus(knex, c.channelId, 100, challengeState);
+
+  const obj = createPendingObjective(c.channelId);
+  await knex.transaction(tx => store.ensureObjective(obj, tx));
+  const response = WalletResponse.initialize();
+
+  await channelDefunder.crank(obj, response);
+
+  const reloadedObjective = await store.getObjective(obj.objectiveId);
+  expect(reloadedObjective.status).toEqual('pending');
+});
+
+test('when the channel is not finalized it defunds the channel by calling pushOutcomeAndWithdraw', async () => {
+  const chainService = new MockChainService();
+  const channelDefunder = ChannelDefunder.create(store, chainService, logger, timingMetrics);
+  const c = channel();
+  const challengeState = {
+    ...c.channelConstants,
+    ...stateVars(),
+  };
+  await Channel.query(knex).withGraphFetched('signingWallet').insert(c);
+  await ChallengeStatus.insertChallengeStatus(knex, c.channelId, 100, challengeState);
+  await ChallengeStatus.setFinalized(knex, c.channelId, 200);
```

Comment:
Why does it take two DB queries to set it as finalized?

(What if the chain service came online, and found a finalized channel?)

Reply:
Originally this was designed **only** with challenges in mind so we could always rely on the `ChallengeRegistered` event occurring first. Oops!

We should probably fix this in #3134